modules/contact/src/main/kotlin/edu/stanford/spezi/modules/contact/model/AddressFormatting.kt

```diff
@@ -0,0 +1,12 @@
+package edu.stanford.spezi.modules.contact.model
+
+import android.location.Address
+
+fun Address.formatted(): String {
```

Comment:
Since we use this inside compose only, we can mark it internal and apply remember here directly:

```kotlin
@Composable
internal fun Address.formatted(): String {
    return remember(this) {
        val lines = (0..maxAddressLineIndex).map { getAddressLine(it) }
        val areaLine = listOf(locality, adminArea, postalCode).mapNotNull { it }.joinToString(" ")
        val countryLine = countryName ?: ""
        ((lines + areaLine) + countryLine)
            .filter { it.isNotBlank() }
            .joinToString("\n")
    }
}
```

Reply:
I would keep the function as is, since I would like to expose it and there it may be used in other contexts, but I will make sure to use the `remember` mechanism whenever I use it, does that sound good? also for the other case

Reply:
Sounds good to me, but let me also share my opinion why i still think it's better given the current usages to use composable. I tend to postpone planning for other usages until whenever those requirements will come.

Your other context currently is testing. There you won't have composable context, hence you wouldn't be able to use this function. Which is still fine, for the test we also want to test this function as well. If you use `formatted()` in compose and `formatted()` in the tests, we won't be able to catch if this implementation would be changed by mistake. In the test I would propose to use the expected value instead

```kotlin
fun assertHasAddress(value: Address?) {
        value?.let {
            val expectedAddressText = "build here the text from the address as done in formatted function. This way we are also testing the "
            address
                .assertExists()
                .onChildAt(0)
                .assertTextEquals(expectedAddressText)
        } ?: address.assertDoesNotExist()
    }
```

Ideally that formatted function should be unit tested - that kind of logic, it is advisable to be handled in a Mapper, which prepares the ui model to be consumed in compose. But since we are only consuming the contact in compose, I think it is still fine to test it via an android test. 

-----

No need for further adjustments for now, but just FYI regarding the simulator approach so that we won't have to repeat building the expected address text in every usage. If we would be using the address card in some other context / composable, we could extract it in a new simulator:

```kotlin
class AddressCardSimulator(
    composeTestRule: ComposeTestRule,
) {
    // this is now an identifier declared in address card
    private val address = composeTestRule.onNodeWithIdentifier(AddressCardTestIdentifier.ADDRESS)

    fun assertHasAddress(value: Address?) {
        value?.let {
            val expectedAddressText = "build here the text from the address as done in formatted function. This way we are also testing the "
            address
                .assertExists()
                .onChildAt(0)
                .assertTextEquals(expectedAddressText)
        } ?: address.assertDoesNotExist()
    }
}
```

And use it inside `ContactComposableSimulator` as follows:

```kotlin
class ContactComposableSimulator(
    private val composeTestRule: ComposeTestRule,
) {
    
    // other nodes and functions...
    
    private val address = AddressCardSimulator(composeTestRule)
    
    fun addressCard(block: AddressCardSimulator.() -> Unit) {
        address.apply(block)
    }
}
```

And contact composable test would look like this

```kotlin
@Test
fun `test displays contact address`() {
    contactComposable {
        addressCard {
            assertHasAddress(contact.address)
        }
    }
}
```